### ISSUE TITLE
Plugin API

### ISSUE THREAD

**Issue by ForbesLindesay** [pugjs/jade-parser#18]
Add an extra option, `plugins` which should be an array of plugin objects.  Plugins can define methods to handle custom token types in any of the places where we have switch statements at the moment.  e.g. if `id` shorthand and `default` from `case`/`when` blocks weren't implemented, here's how you could implement them in a plugin:

```js
opts = {
  plugins: [
    {
      expressionTokens: {
        'id': function (parser) {
          parser.tokens.defer({
            type: 'tag',
            val: 'div',
            line: parser.peek().line,
            filename: parser.filename
          });
          return parser.parseExpr();
        }
      },
      tagAttributeTokens: {
        'id': function (parser, tag) {
          var tok = parser.expect('id');
          tag.attrs.push({
            name: tok.type,
            val: "'" + tok.val + "'",
            escaped: false
          });
        }
      },
      caseTokens: {
        'default': function (parser, block) {
          var tok = parser.expect('default');
          block.nodes.push({
            type: 'When',
            expr: 'default',
            block: parser.parseBlockExpansion(),
            debug: false,
            line: tok.line,
            filename: parser.filename
          });
        }
      }
    }
  ]
};
```

N.B. we should throw an exception if two plugins attempt to handle the same token type in the same place.  We should also probably not allow plugins to handle built in tokens here (built in tokens can be handled in a post-parse plugin that can walk the tree and re-write the output AST or by re-writing tokens to a different name in a pre-parse plugin).

**Comment by TimothyGu** [comment]
:+1:

> built in tokens can be handled in a post-parse plugin that can walk the tree and re-write the output AST or by re-writing tokens to a different name in a pre-parse plugin

We should also encourage plugin writers to properly namespace their token types if the plugin is a full suite of lexer and parser plugins so that there is no conflict.

**Comment by TimothyGu** [comment]
Implemented in #22. I deviated from your implementation a bit: all void-returning functions now need to return `true` to tell the parser that there is a plugin available to do the work correctly.

So the final, working plugin set is:

```js
opts = {
  plugins: [
    {
      expressionTokens: {
        'id': function (parser) {
          parser.tokens.defer({
            type: 'tag',
            val: 'div',
            line: parser.peek().line,
            filename: parser.filename
          });
          return parser.parseExpr();
        }
      },
      tagAttributeTokens: {
        'id': function (parser, tag) {
          var tok = parser.expect('id');
          tag.attrs.push({
            name: tok.type,
            val: "'" + tok.val + "'",
            mustEscape: false
          });
          return true;
        }
      },
      caseTokens: {
        'default': function (parser, block) {
          var tok = parser.expect('default');
          block.nodes.push({
            type: 'When',
            expr: 'default',
            block: parser.parseBlockExpansion(),
            debug: false,
            line: tok.line,
            filename: parser.filename
          });
          return true;
        }
      }
    }
  ]
};
```

**Comment by uinz** [comment]
hello, i want get way to handle class by my self
that i want use module css in the server side

https://github.com/pugjs/pug/issues/2474
